### PR TITLE
gh-3332 MySQL: Fix SQL query for deleting Batch Step Execution Context data

### LINKED ISSUE #3332: MySQL: Fix SQL query for deleting Batch Step Execution Context data

### PR DESCRIPTION
**Pull request by ghillert**
resolves #3332

### CODE REVIEW

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcDataflowJobExecutionDao.java [hunk 1]
```diff
@@ -42,13 +42,9 @@
...
 	 */
 	private static final String  SQL_DELETE_BATCH_STEP_EXECUTION_CONTEXT =
-			"DELETE FROM %PREFIX%STEP_EXECUTION_CONTEXT " +
-			"WHERE STEP_EXECUTION_ID " +
-			"IN ( " +
-			"SELECT SEC.STEP_EXECUTION_ID " +
-			"FROM %PREFIX%STEP_EXECUTION_CONTEXT SEC " +
-			"JOIN %PREFIX%STEP_EXECUTION SE ON SE.STEP_EXECUTION_ID = SEC.STEP_EXECUTION_ID " +
-			"WHERE SE.JOB_EXECUTION_ID in (:jobExecutionIds))";
+			"DELETE SEC.* FROM BATCH_STEP_EXECUTION_CONTEXT SEC " +
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcDataflowJobExecutionDao.java]: @ghillert Could you please some description in the commit to better understand what went wrong and what is fixed with this PR? Thanks!

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcDataflowJobExecutionDao.java [hunk 2]
```diff
@@ -38,17 +44,27 @@
 	private final NamedParameterJdbcTemplate jdbcTemplate;
 	private final String tablePrefix;
 
+	//	/**
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcDataflowJobExecutionDao.java]: This comment can be removed

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcDataflowJobExecutionDao.java [hunk 3]
```diff
@@ -38,17 +44,27 @@
...
 	 */
 	private static final String  SQL_DELETE_BATCH_STEP_EXECUTION_CONTEXT =
 			"DELETE FROM %PREFIX%STEP_EXECUTION_CONTEXT " +
 			"WHERE STEP_EXECUTION_ID " +
-			"IN ( " +
-			"SELECT SEC.STEP_EXECUTION_ID " +
-			"FROM %PREFIX%STEP_EXECUTION_CONTEXT SEC " +
-			"JOIN %PREFIX%STEP_EXECUTION SE ON SE.STEP_EXECUTION_ID = SEC.STEP_EXECUTION_ID " +
-			"WHERE SE.JOB_EXECUTION_ID in (:jobExecutionIds))";
+			"IN (:stepExecutionIds)";
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcDataflowJobExecutionDao.java]: Will we have issues with step execution size in this in clause.   I've seen jobs with hundreds of steps.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskDeleteService.java [hunk 4]
```diff
@@ -216,7 +216,16 @@ public void deleteTaskExecutions(Set<Long> taskExecutionIds) {
 		auditData.put("Deleted Job Execution IDs", StringUtils.collectionToDelimitedString(jobExecutionIds, ", "));
 
 		if (!jobExecutionIds.isEmpty()) {
-			final int numberOfDeletedBatchStepExecutionContextRows = dataflowJobExecutionDao.deleteBatchStepExecutionContextByJobExecutionIds(jobExecutionIds);
+			final Set<Long> stepExecutionIds = dataflowJobExecutionDao.findStepExecutionIds(jobExecutionIds);
+
+			final int numberOfDeletedBatchStepExecutionContextRows;
+			if (!stepExecutionIds.isEmpty()) {
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskDeleteService.java]: Why not use `stepExecutionIds.size()`?

**Comment by cppwfs** [comment]
As discussed in the slack channel, mysql is failing to do the delete:
Caused by: java.sql.SQLException: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'BJI WHERE NOT EXISTS ( SELECT JOB_INSTANCE_ID FROM BATCH_JOB_EXECUTION BJE WHERE' at line 1

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcDataflowJobExecutionDao.java (hunk 2 above)
Comment (ghillert) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcDataflowJobExecutionDao.java]: Removed.

**Comment by cppwfs** [comment]
Change works on DB2, mysql 5.7 & 8.x.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskDeleteService.java (hunk 4 above)
Comment (ghillert) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskDeleteService.java]: With `size()` I would need to check `stepExecutionIds.size() > 0`. I usually prefer the `!stepExecutionIds.isEmpty()` notation. If you feel strongly I can change.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcDataflowJobExecutionDao.java (hunk 3 above)
Comment (ghillert) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcDataflowJobExecutionDao.java]: Created  https://github.com/spring-cloud/spring-cloud-dataflow/issues/3348

**Comment by sabbyanandan** [comment]
@cppwfs: Is this good to go then?

**Comment by cppwfs** [comment]
Doing a full build to re-verify tests and to make sure checkstyle enforcement rules are followed.

**Comment by cppwfs** [comment]
LGTM.

**Comment by cppwfs** [comment]
Rebased, Squashed, Merged.

**Pull request closed without merging**